File: src/python/xerxes/memory/entity_memory.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Entity-centric memory tier with lightweight relationship tracking.

``EntityMemory`` extracts noun-phrase and quoted entities from saved
content, records first/last-seen timestamps and mention counts per
entity, and infers a handful of common relationships (works_with,
knows, created, ...) via regex patterns. It exposes graph-style
queries like ``get_entity_info`` and ``get_related_entities``."""

import re
from collections import defaultdict
from typing import Any

from .base import Memory, MemoryItem


class EntityMemory(Memory):
    """Memory tier that indexes content by extracted entities and relations."""

    def __init__(
        self,
        storage: Any | None = None,
        max_items: int = 5000,
        enable_embeddings: bool = False,
    ) -> None:
        """Initialise the entity tier and its auxiliary indexes.

        Args:
            storage: Optional ``MemoryStorage`` for durable persistence.
            max_items: Soft cap on retained items.
            enable_embeddings: Whether to compute embeddings on save."""

        super().__init__(storage=storage, max_items=max_items, enable_embeddings=enable_embeddings)
        self.entities: dict[str, dict[str, Any]] = {}
        self.relationships: dict[str, list[tuple[str, str]]] = defaultdict(list)
        self.entity_mentions: dict[str, list[str]] = defaultdict(list)

    def save(
        self, content: str, metadata: dict[str, Any] | None = None, entities: list[str] | None = None, **kwargs
    ) -> MemoryItem:
        """Save ``content`` and index it under the supplied or extracted entities.

        Args:
            content: Text payload to remember.
            metadata: Annotations; receives an ``entities`` field.
            entities: Pre-extracted entity list; when omitted, entities
                are inferred from ``content`` via ``_extract_entities``.
            **kwargs: Accepted for protocol compatibility; unused."""

        metadata = metadata or {}

        if not entities:
            entities = self._extract_entities(content)

        metadata["entities"] = entities

        item = MemoryItem(
            content=content,
            memory_type="entity",
            metadata=metadata,
        )

        for entity in entities:
            self._update_entity(entity, item)

        relationships = self._extract_relationships(content, entities)
        for entity1, relation, entity2 in relationships:
            self.relationships[relation].append((entity1, entity2))

        self._items.append(item)
        self._index[item.memory_id] = item

        if self.storage:
            self.storage.save(f"entity_{item.memory_id}", item.to_dict())
            self._save_entity_data()

        return item

    def search(
        self,
        query: str,
        limit: int = 10,
        filters: dict[str, Any] | None = None,
        entity_filter: list[str] | None = None,
        **kwargs,
    ) -> list[MemoryItem]:
        """Return items whose indexed entities overlap with the query's.

        Each candidate item's ``relevance_score`` is set to the overlap
        ratio against the target entity set before sorting.

        Args:
            query: Text whose entities are extracted as the target set
                unless ``entity_filter`` is supplied.
            limit: Maximum results.
            filters: Attribute-level filters applied to items.
            entity_filter: Explicit entity allow-list overriding query
                extraction."""

        query_entities = self._extract_entities(query)
        target_entities = entity_filter or query_entities

        matches = []

        for item in self._items:
            item_entities = item.metadata.get("entities", [])

            if target_entities:
                overlap = set(item_entities) & set(target_entities)
                if not overlap:
                    continue

            if filters:
                skip = False
                for key, value in filters.items():
                    if hasattr(item, key) and getattr(item, key) != value:
                        skip = True
                        break
                if skip:
                    continue

            if target_entities:
                item.relevance_score = len(overlap) / len(target_entities)
            else:
                item.relevance_score = 1.0 if query.lower() in item.content.lower() else 0.5

            matches.append(item)

        matches.sort(key=lambda x: x.relevance_score, reverse=True)
        return matches[:limit]

    def retrieve(
        self,
        memory_id: str | None = None,
        filters: dict[str, Any] | None = None,
        limit: int = 10,
    ) -> MemoryItem | list[MemoryItem] | None:
        """Fetch by id, or scan items for the first ``limit`` matching ``filters``."""

        if memory_id:
            return self._index.get(memory_id)

        results = []
        for item in self._items:
            if filters:
                skip = False
                for key, value in filters.items():
                    if hasattr(item, key) and getattr(item, key) != value:
                        skip = True
                        break
                if skip:
                    continue

            results.append(item)
            if len(results) >= limit:
                break

        return results

    def update(self, memory_id: str, updates: dict[str, Any]) -> bool:
        """Apply ``updates`` to the item, re-indexing entities when content changes."""

        if memory_id not in self._index:
            return False

        item = self._index[memory_id]

        if "content" in updates:
            old_entities = item.metadata.get("entities", [])
            new_entities = self._extract_entities(updates["content"])
            updates.setdefault("metadata", {})["entities"] = new_entities

            for entity in old_entities:
                if entity in self.entity_mentions:
                    self.entity_mentions[entity].remove(memory_id)

            for entity in new_entities:
                self.entity_mentions[entity].append(memory_id)

        for key, value in updates.items():
            if hasattr(item, key):
                setattr(item, key, value)

        if self.storage:
            self.storage.save(f"entity_{memory_id}", item.to_dict())
            self._save_entity_data()

        return True

    def delete(self, memory_id: str | None = None, filters: dict[str, Any] | None = None) -> int:
        """Remove the item with ``memory_id`` and unlink its entity mentions.

        Bulk filter-based deletion is not implemented; ``filters`` is
        accepted for protocol compatibility but ignored. Returns 1 on
        successful single delete, 0 otherwise."""

        count = 0

        if memory_id and memory_id in self._index:
            item = self._index[memory_id]

            for entity in item.metadata.get("entities", []):
                if entity in self.entity_mentions:
                    self.entity_mentions[entity].remove(memory_id)

            self._items.remove(item)
            del self._index[memory_id]
            if self.storage:
                self.storage.delete(f"entity_{memory_id}")
            count = 1

        return count

    def clear(self) -> None:
        """Drop every item, entity, relationship, and mention; purge storage entries."""

        self._items.clear()
        self._index.clear()
        self.entities.clear()
        self.relationships.clear()
        self.entity_mentions.clear()

        if self.storage:
            for key in self.storage.list_keys("entity_"):
                self.storage.delete(key)

    def get_entity_info(self, entity: str) -> dict[str, Any]:
        """Return a snapshot of ``entity``'s tracked metadata, mentions, and relations."""

        info = self.entities.get(entity, {})
        info["mentions"] = self.entity_mentions.get(entity, [])
        info["relationships"] = []

        for relation, pairs in self.relationships.items():
            for e1, e2 in pairs:
                if e1 == entity:
                    info["relationships"].append({"relation": relation, "target": e2})
                elif e2 == entity:
                    info["relationships"].append({"relation": f"inverse_{relation}", "target": e1})

        return info

    def get_related_entities(self, entity: str, max_depth: int = 2) -> set[str]:
        """Walk the relationship graph from ``entity`` up to ``max_depth`` hops.

        Returns the set of entities reachable through any relationship
        (in either direction), excluding ``entity`` itself."""

        related = set()
        to_explore = [(entity, 0)]
        explored = set()

        while to_explore:
            current, depth = to_explore.pop(0)
            if current in explored or depth > max_depth:
                continue

            explored.add(current)

            for _relation, pairs in self.relationships.items():
                for e1, e2 in pairs:
                    if e1 == current:
                        related.add(e2)
                        if depth < max_depth:
                            to_explore.append((e2, depth + 1))
                    elif e2 == current:
                        related.add(e1)
                        if depth < max_depth:
                            to_explore.append((e1, depth + 1))

        return related

    def _extract_entities(self, text: str) -> list[str]:
        """Heuristic entity extraction: capitalised phrases + quoted strings."""

        entities = []

        pattern = r"\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b"
        matches = re.findall(pattern, text)
        entities.extend(matches)

        quoted = re.findall(r'"([^"]*)"', text)
        entities.extend(quoted)

        common_words = {"The", "This", "That", "These", "Those"}
        entities = list(set(e for e in entities if e not in common_words))

        return entities

    def _extract_relationships(self, text: str, entities: list[str]) -> list[tuple[str, str, str]]:
        """Match a handful of regex relationship templates over ``text``.

        Returns ``(subject, relation, object)`` triples whose subject
        and object both appear in ``entities``."""

        relationships = []

        patterns = [
            (r"(\w+)\s+is\s+(?:a|an|the)?\s*(\w+)\s+of\s+(\w+)", "relation_of"),
            (r"(\w+)\s+works\s+(?:at|for|with)\s+(\w+)", "works_with"),
            (r"(\w+)\s+knows\s+(\w+)", "knows"),
            (r"(\w+)\s+created\s+(\w+)", "created"),
        ]

        for pattern, relation in patterns:
            matches = re.finditer(pattern, text, re.IGNORECASE)
            for match in matches:
                groups = match.groups()
                if len(groups) >= 2:
                    e1, e2 = groups[0], groups[-1]
                    if e1 in entities and e2 in entities:
                        relationships.append((e1, relation, e2))

        return relationships

    def _update_entity(self, entity: str, memory_item: MemoryItem) -> None:
        """Increment the entity's frequency and append a context snippet."""

        if entity not in self.entities:
            self.entities[entity] = {"first_seen": memory_item.timestamp, "frequency": 0, "contexts": []}

        self.entities[entity]["frequency"] += 1
        self.entities[entity]["last_seen"] = memory_item.timestamp
        self.entities[entity]["contexts"].append(memory_item.content[:100])
        self.entity_mentions[entity].append(memory_item.memory_id)

    def _save_entity_data(self) -> None:
        """Persist the entity/relationship/mention tables to storage."""

        if self.storage:
            self.storage.save("_entity_entities", self.entities)
            self.storage.save("_entity_relationships", dict(self.relationships))
            self.storage.save("_entity_mentions", dict(self.entity_mentions))

File: src/typescript/src/memory/entityMemory.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  Memory,
  MemoryItem,
  type MemoryFilters,
  type MemoryMetadata,
  type MemorySaveOptions,
  type MemorySearchOptions,
  type MemoryUpdate,
} from './base.js'
import type { MemoryStorage } from './storage.js'

const COMMON_ENTITY_WORDS = new Set(['The', 'This', 'That', 'These', 'Those'])
const ENTITY_PHRASE_PATTERN = /\b[A-Z][a-z]+(?:\s+[A-Z][a-z]+)*\b/g
const QUOTED_ENTITY_PATTERN = /"([^"]*)"/g

const RELATIONSHIP_PATTERNS: ReadonlyArray<readonly [RegExp, string]> = [
  [/(\w+)\s+is\s+(?:a|an|the)?\s*(\w+)\s+of\s+(\w+)/gi, 'relation_of'],
  [/(\w+)\s+works\s+(?:at|for|with)\s+(\w+)/gi, 'works_with'],
  [/(\w+)\s+knows\s+(\w+)/gi, 'knows'],
  [/(\w+)\s+created\s+(\w+)/gi, 'created'],
]

export interface EntityMemoryOptions {
  readonly enableEmbeddings?: boolean
  readonly maxItems?: number
  readonly storage?: MemoryStorage
}

export interface EntitySaveOptions extends MemorySaveOptions {
  readonly entities?: readonly string[]
}

export interface EntitySearchOptions extends MemorySearchOptions {
  readonly entityFilter?: readonly string[]
}

export interface EntityRecord {
  readonly contexts: string[]
  readonly firstSeen: Date
  frequency: number
  lastSeen: Date
}

export interface EntityRelationship {
  readonly relation: string
  readonly target: string
}

export interface EntityInfo {
  readonly contexts?: readonly string[]
  readonly firstSeen?: Date
  readonly frequency?: number
  readonly lastSeen?: Date
  readonly mentions: readonly string[]
  readonly relationships: readonly EntityRelationship[]
}

type EntityRelationPair = readonly [source: string, target: string]

/**
 * A memory tier indexed by lightweight named-entity and relationship heuristics.
 *
 * It intentionally keeps the extraction rule deterministic and dependency-free:
 * capitalized phrases and quoted text become entities, while a small set of
 * relation phrases build a traversable graph.
 */
export class EntityMemory extends Memory {
  readonly entities: Record<string, EntityRecord> = {}
  readonly entityMentions: Record<string, string[]> = {}
  readonly relationships: Record<string, EntityRelationPair[]> = {}

  constructor(options: EntityMemoryOptions = {}) {
    super(options.storage, options.maxItems ?? 5_000, options.enableEmbeddings ?? false)
  }

  clear(): void {
    this.items.length = 0
    this.index.clear()
    clearRecord(this.entities)
    clearRecord(this.relationships)
    clearRecord(this.entityMentions)

    for (const key of this.storage?.listKeys('entity_') ?? []) this.storage?.delete(key)
  }

  delete(memoryId?: string, _filters?: MemoryFilters): number {
    if (!memoryId) return 0
    const item = this.index.get(memoryId)
    if (!item) return 0

    for (const entity of entityNames(item)) removeMention(this.entityMentions, entity, memoryId)
    this.remove(item)
    this.storage?.delete(entityStorageKey(memoryId))
    return 1
  }

  /** Return entities inferred from capitalized phrases and double-quoted text. */
  extractEntities(text: string): string[] {
    const entities = new Set<string>()
    for (const match of text.matchAll(ENTITY_PHRASE_PATTERN)) {
      const entity = match[0]
      if (entity && !COMMON_ENTITY_WORDS.has(entity)) entities.add(entity)
    }
    for (const match of text.matchAll(QUOTED_ENTITY_PATTERN)) {
      const entity = match[1]
      if (entity && !COMMON_ENTITY_WORDS.has(entity)) entities.add(entity)
    }
    return [...entities]
  }

  /** Infer directed relationship triples where both ends are known entities. */
  extractRelationships(text: string, entities: readonly string[]): Array<readonly [string, string, string]> {
    const known = new Set(entities)
    const relationships: Array<readonly [string, string, string]> = []
    for (const [pattern, relation] of RELATIONSHIP_PATTERNS) {
      pattern.lastIndex = 0
      for (const match of text.matchAll(pattern)) {
        const source = match[1]
        const target = match.at(-1)
        if (source && target && known.has(source) && known.has(target)) {
          relationships.push([source, relation, target])
        }
      }
    }
    return relationships
  }

  /** Return tracked metadata, mentions, and incoming/outgoing relation edges. */
  getEntityInfo(entity: string): EntityInfo {
    const record = this.entities[entity]
    const relationships: EntityRelationship[] = []
    for (const [relation, pairs] of Object.entries(this.relationships)) {
      for (const [source, target] of pairs) {
        if (source === entity) relationships.push({ relation, target })
        else if (target === entity) relationships.push({ relation: `inverse_${relation}`, target: source })
      }
    }
    return {
      ...(record ? {
        contexts: [...record.contexts],
        firstSeen: record.firstSeen,
        frequency: record.frequency,
        lastSeen: record.lastSeen,
      } : {}),
      mentions: [...(this.entityMentions[entity] ?? [])],
      relationships,
    }
  }

  /** Walk relationship edges in either direction, excluding the source entity. */
  getRelatedEntities(entity: string, maxDepth = 2): Set<string> {
    const related = new Set<string>()
    const toExplore: Array<readonly [string, number]> = [[entity, 0]]
    const explored = new Set<string>()

    while (toExplore.length > 0) {
      const next = toExplore.shift()
      if (!next) continue
      const [current, depth] = next
      if (explored.has(current) || depth > maxDepth) continue
      explored.add(current)

      for (const pairs of Object.values(this.relationships)) {
        for (const [source, target] of pairs) {
          if (source === current) {
            related.add(target)
            if (depth < maxDepth) toExplore.push([target, depth + 1])
          } else if (target === current) {
            related.add(source)
            if (depth < maxDepth) toExplore.push([source, depth + 1])
          }
        }
      }
    }
    related.delete(entity)
    return related
  }

  retrieve(memoryId?: string, filters?: MemoryFilters, limit = 10): MemoryItem | MemoryItem[] | undefined {
    if (memoryId) return this.index.get(memoryId)
    return this.items.filter(item => this.matchesFilters(item, filters)).slice(0, limit)
  }

  save(content: string, metadata: MemoryMetadata = {}, options: EntitySaveOptions = {}): MemoryItem {
    const entities = options.entities?.length ? [...options.entities] : this.extractEntities(content)
    const item = new MemoryItem({
      content,
      memoryType: 'entity',
      metadata: { ...metadata, entities },
      ...(options.agentId ? { agentId: options.agentId } : {}),
      ...(options.taskId ? { taskId: options.taskId } : {}),
      ...(options.userId ? { userId: options.userId } : {}),
      ...(options.conversationId ? { conversationId: options.conversationId } : {}),
    })

    for (const entity of entities) this.updateEntity(entity, item)
    for (const [source, relation, target] of this.extractRelationships(content, entities)) {
      const pairs = this.relationships[relation] ?? []
      pairs.push([source, target])
      this.relationships[relation] = pairs
    }

    this.append(item)
    this.persist(item)
    return item
  }

  search(query: string, limit = 10, filters?: MemoryFilters, options: EntitySearchOptions = {}): MemoryItem[] {
    const queryEntities = this.extractEntities(query)
    const targetEntities = options.entityFilter?.length ? options.entityFilter : queryEntities
    const target = new Set(targetEntities)
    const matches: MemoryItem[] = []

    for (const item of this.items) {
      if (!this.matchesFilters(item, filters)) continue
      const itemEntities = entityNames(item)
      const overlap = itemEntities.filter(entity => target.has(entity))
      if (target.size > 0 && overlap.length === 0) continue
      item.relevanceScore = target.size > 0
        ? overlap.length / target.size
        : item.content.toLowerCase().includes(query.toLowerCase()) ? 1 : 0.5
      matches.push(item)
    }
    return matches.sort((left, right) => right.relevanceScore - left.relevanceScore).slice(0, limit)
  }

  update(memoryId: string, updates: MemoryUpdate): boolean {
    const item = this.index.get(memoryId)
    if (!item) return false

    if (updates.content !== undefined) {
      const oldEntities = entityNames(item)
      const newEntities = this.extractEntities(updates.content)
      for (const entity of oldEntities) removeMention(this.entityMentions, entity, memoryId)
      for (const entity of newEntities) {
        const mentions = this.entityMentions[entity] ?? []
        mentions.push(memoryId)
        this.entityMentions[entity] = mentions
      }
      this.updateItem(item, {
        ...updates,
        metadata: { ...item.metadata, ...(updates.metadata ?? {}), entities: newEntities },
      })
    } else {
      this.updateItem(item, updates)
    }

    this.persist(item)
    return true
  }

  private persist(item: MemoryItem): void {
    if (!this.storage) return
    this.storage.save(entityStorageKey(item.memoryId), item.toRecord())
    this.storage.save('_entity_entities', this.entities)
    this.storage.save('_entity_relationships', this.relationships)
    this.storage.save('_entity_mentions', this.entityMentions)
  }

  private updateEntity(entity: string, item: MemoryItem): void {
    const existing = this.entities[entity]
    if (existing) {
      existing.frequency += 1
      existing.lastSeen = item.timestamp
      existing.contexts.push(item.content.slice(0, 100))
    } else {
      this.entities[entity] = {
        firstSeen: item.timestamp,
        frequency: 1,
        lastSeen: item.timestamp,
        contexts: [item.content.slice(0, 100)],
      }
    }
    const mentions = this.entityMentions[entity] ?? []
    mentions.push(item.memoryId)
    this.entityMentions[entity] = mentions
  }
}

function clearRecord(record: Record<string, unknown>): void {
  for (const key of Object.keys(record)) delete record[key]
}

function entityNames(item: MemoryItem): string[] {
  const entities = item.metadata.entities
  return Array.isArray(entities) ? entities.filter((entity): entity is string => typeof entity === 'string') : []
}

function entityStorageKey(memoryId: string): string {
  return `entity_${memoryId}`
}

function removeMention(mentions: Record<string, string[]>, entity: string, memoryId: string): void {
  const values = mentions[entity]
  if (!values) return
  const position = values.indexOf(memoryId)
  if (position >= 0) values.splice(position, 1)
}
